grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcDecoratingService.java

```diff
@@ -68,29 +69,39 @@ public void serviceAdded(ServiceConfig cfg) throws Exception {
                                                          .config()
                                                          .dependencyInjector();
 
-        final Map<ServerMethodDefinition<?, ?>, List<DecoratorAndOrder>> registryDecorators =
-                handlerRegistry.decorators();
-
-        final Builder<ServerMethodDefinition<?, ?>, HttpService> builder = ImmutableMap.builder();
+        final Map<ServerMethodDefinition<?, ?>, HttpService> decorated = new HashMap<>();
 
+        final Map<ServerMethodDefinition<?, ?>, List<DecoratorAndOrder>> annotationDecorators =
+                handlerRegistry.annotationDecorators();
         for (Entry<ServerMethodDefinition<?, ?>, List<DecoratorAndOrder>> entry
-                : registryDecorators.entrySet()) {
+                : annotationDecorators.entrySet()) {
             final List<? extends Function<? super HttpService, ? extends HttpService>> decorators =
                     entry.getValue()
                          .stream()
                          .map(decoratorAndOrder -> decoratorAndOrder.decorator(dependencyInjector))
                          .collect(toImmutableList());
-            builder.put(entry.getKey(), applyDecorators(decorators, delegate));
+            decorated.put(entry.getKey(), applyDecorators(decorators, delegate));
+        }
+
+        final Map<ServerMethodDefinition<?, ?>, Iterable<? extends Function<? super HttpService,
+                ? extends HttpService>>> additionalDecorators = handlerRegistry.additionalDecorators();
+        for (Entry<ServerMethodDefinition<?, ?>, Iterable<? extends Function<? super HttpService,
+                ? extends HttpService>>> entry : additionalDecorators.entrySet()) {
+            final HttpService service = decorated.getOrDefault(entry.getKey(), delegate);
+            decorated.put(entry.getKey(), applyDecorators(entry.getValue(), service));
         }
-        decorated = builder.build();
```

Comment:
How about hiding `additionalDecorators` from GrpcDecoratingService.java and delegating `handlerRegistry` to return a combined list of decorators like:
```java
Map<ServerMethodDefinition<?, ?>, Iterable<? extends Function<? super HttpService,
                ? extends HttpService>>> decorators = 
    handlerRegistry.decorators(dependencyInjector);
```

Reply:
That's a brilliant idea. 👍